frame/contracts/proc-macro/src/lib.rs

```diff
@@ -606,6 +607,53 @@ fn expand_functions(def: &EnvDef, expand_blocks: bool, host_state: TokenStream2)
 		let is_stable = f.is_stable;
 		let not_deprecated = f.not_deprecated;
 
+		// wrapped host function body call with host function traces
+		// see https://github.com/paritytech/substrate/tree/master/frame/contracts#host-function-tracing
+		let wrapped_body_with_trace = {
+			let params_fmt_str = params.clone().filter_map(|arg| match arg {
+				syn::FnArg::Receiver(_) => None,
+				syn::FnArg::Typed(p) => {
+					match *p.pat.clone() {
+						syn::Pat::Ident(ref pat_ident) => Some(pat_ident.ident.to_string()),
+						_ => None,
+					}
+				},
+			})
+			.map(|s| format!("{s}: {{:?}}"))
+			.collect::<Vec<_>>()
+			.join(", ");
+
+			let trace_fmt_str = format!("{}::{}({}) = {{:?}}", module, name, params_fmt_str);
+			let trace_fmt_args = params.clone().filter_map(|arg| match arg {
+				syn::FnArg::Receiver(_) => None,
+				syn::FnArg::Typed(p) => {
+					match *p.pat.clone() {
+						syn::Pat::Ident(ref pat_ident) => Some(pat_ident.ident.clone()),
+						_ => None,
+					}
+				},
+			});
+
+			let debug_buffer_fmt_str = format!("{}\n", trace_fmt_str);
+			let debug_buffer_fmt_args = trace_fmt_args.clone();
```

Comment:
just a little DRYing refactoring suggestion
```suggestion
		// Wrapped host function body call with host function traces.
		// See https://github.com/paritytech/substrate/tree/master/frame/contracts#host-function-tracing
		let wrapped_body_with_trace = {
			let params_filter_map = |arg: &syn::FnArg| match arg {
				syn::FnArg::Receiver(_) => None,
				syn::FnArg::Typed(p) => {
					match *p.pat.clone() {
						syn::Pat::Ident(ref pat_ident) => Some(pat_ident.ident.clone()),
						_ => None,
					}
				},
			};

			let params_fmt_str = params.clone().filter_map(params_filter_map)
			.map(|s| format!("{}: {{:?}}", s.to_string()))
			.collect::<Vec<_>>()
			.join(", ");
			let trace_fmt_str = format!("{}::{}({}) = {{:?}}", module, name, params_fmt_str);
			let trace_fmt_args = params.clone().filter_map(params_filter_map);
			let debug_buffer_fmt_str = format!("{}\n", trace_fmt_str);
			let debug_buffer_fmt_args = trace_fmt_args.clone();
```

Reply:
ah nice good catch, I think I can just clone the mapped iterator, instead of creating a closure